File: defis/ParcoursDeMiel/ParcoursDeMiel.py
import math

points = []
dist = []
n = 0

def read_points(filename):
    points = []
    with open(filename, "r") as f:
        for line in f:
            line = line.strip()
            if not line:
                continue
            x, y = map(float, line.split(","))
            points.append((x, y))
    return points

def distance_matrix(points):
    matrix = [[0.0] * n for _ in range(n)]

    for i in range(n):
        x1, y1 = points[i]
        for j in range(n):
            x2, y2 = points[j]
            matrix[i][j] = math.hypot(x2 - x1, y2 - y1)

    return matrix

def initialize(filename):
    global dist
    global n
    global points
    points = read_points(filename)
    #print(points)
    n = len(points)
    dist = distance_matrix(points)
    #print(dist)

def calculate_length(parcours):
    if parcours == None :
        return -1
    total_distance = 0
    num_points = len(parcours)
    
    for i in range(num_points):
        u = parcours[i]
        v = parcours[(i + 1) % num_points]
        total_distance += dist[u][v]
        
    return total_distance


def advise_honeybee():

    # *** recommandation : écrire votre code ici ***

    return None

if __name__ == "__main__":
    initialize("input_demo.txt")
    parcours_exemple = [i for i in range(n)]
    length = calculate_length(parcours_exemple)
    print(parcours_exemple)
    print(length)

    # pour le input_demo.txt, le meilleur parcours est [1, 0, 6, 5, 3, 2, 4]
    # 27.699768073747965
    #parcours_exemple = [1, 0, 6, 5, 3, 2, 4]

    best_answer = advise_honeybee()
    print(best_answer)
    print(calculate_length(best_answer))

File: defis/ParcoursDeMiel/ParcoursDeMiel.java
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

public class ParcoursDeMiel {

    private static List<Point> points = new ArrayList<>();
    private static double[][] dist;
    private static int n = 0;

    private static final class Point {
        final double x;
        final double y;
        Point(double x, double y) { this.x = x; this.y = y; }
    }

    private static List<Point> readPoints(String filename) throws IOException {
        List<Point> result = new ArrayList<>();
        for (String line : Files.readAllLines(Paths.get(filename))) {
            String trimmed = line.trim();
            if (trimmed.isEmpty()) continue;
            String[] parts = trimmed.split(",");
            double x = Double.parseDouble(parts[0].trim());
            double y = Double.parseDouble(parts[1].trim());
            result.add(new Point(x, y));
        }
        return result;
    }

    private static double[][] distanceMatrix(List<Point> pts) {
        int size = pts.size();
        double[][] matrix = new double[size][size];
        for (int i = 0; i < size; i++) {
            Point a = pts.get(i);
            for (int j = 0; j < size; j++) {
                Point b = pts.get(j);
                matrix[i][j] = Math.hypot(b.x - a.x, b.y - a.y);
            }
        }
        return matrix;
    }

    private static void initialize(String filename) throws IOException {
        points = readPoints(filename);
        n = points.size();
        dist = distanceMatrix(points);
    }

    private static double calculateLength(List<Integer> parcours) {
        if (parcours == null) return -1.0;
        double totalDistance = 0.0;
        int numPoints = parcours.size();
        for (int i = 0; i < numPoints; i++) {
            int u = parcours.get(i);
            int v = parcours.get((i + 1) % numPoints);
            totalDistance += dist[u][v];
        }
        return totalDistance;
    }

    private static List<Integer> adviseHoneybee() {
        
        // *** recommandation : écrire votre code ici ***

        return null;
    }

    public static void main(String[] args) throws IOException {
        initialize("input_demo.txt");

        // parcours_exemple = [i for i in range(n)]
        List<Integer> parcoursExemple = IntStream.range(0, n).boxed().collect(Collectors.toList());
        double length = calculateLength(parcoursExemple);
        System.out.println(parcoursExemple);
        System.out.println(length);

        // For input_demo.txt, the best route is [1, 0, 6, 5, 3, 2, 4]
        // 27.699768073747965
        // parcours_exemple = [1, 0, 6, 5, 3, 2, 4]

        List<Integer> bestAnswer = adviseHoneybee();
        System.out.println(bestAnswer);
        System.out.println(calculateLength(bestAnswer));
    }
}
